File: 0015-3sum/0015-3sum.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def three_sum(nums)
 
    nums.sort!
    output=[]
    for i in 0..nums.length-3
        next if i>0&&nums[i-1]==nums[i]
        j,k=i+1,nums.length-1
        while j<k
            if j-1!=i&&nums[j-1]==nums[j]
                j+=1
                next
            end
            if k+1<nums.length&&nums[k+1]==nums[k]
                k-=1
                next
            end
            sum=nums[i]+nums[j]+nums[k]
            if sum==0
                output.push([nums[i],nums[j],nums[k]])
                j+=1
            elsif sum>0
                k-=1
            else
                j+=1
            end
        end
    end
        output
end

File: 0015-3sum/0015-3sum.ts
function threeSum(nums: number[]): number[][] {
    let sorted = nums.sort((a,b)=>a-b);
    const result = [];
    // console.log(sorted)
    for(let i=0; i<sorted.length-2; i++){
        let l=i+1, r=sorted.length-1;
        while(l<r){
            let sum = sorted[l]+sorted[i]+sorted[r];

            if(sum==0)
                result.push([sorted[l],sorted[i],sorted[r]])
            if(sum>0){
                    let nr=sorted[r]
                do {
                    r--
                } while(r>l && sorted[r]==nr);
                    
            }
            else {
                   let nl=sorted[l]
                do {
                    l++
                } while(l<r && sorted[l]==nl);
            }
        }
        while(i<sorted.length-1 && sorted[i]==sorted[i+1])
            i++
    }
    return result;
};